core/lib/zksync_core/src/api_server/web3/namespaces/eth.rs

```diff
@@ -826,6 +826,10 @@ impl<G: L1GasPriceProvider> EthNamespace<G> {
                     .state
                     .resolve_filter_block_number(filter.to_block)
                     .await?;
+                let latest_block = self
```

Comment:
- We have a similar TOCTOU issue here as with `to_block`; `latest_block` may get outdated before events are fetched. If we want a quick fix, I guess the initial PR version would do better (although, as we've discussed, it's not very efficient with DB queries). If we want to avoid TOCTOU, I think we'd need to rework getting events from DB so that it relies on specific block ranges; or we could issue all queries in a transaction. (Although the transaction-based solution is simpler, IMO, in this case simplifying DAL logic would be warranted.)
- As an optimization, we can avoid this query if `filter.to_block` is not set or set to "latest" (which I suspect would be a common occurrence).